x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1149,22 +1150,48 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
         @Override
         protected LogicalPlan rule(LogicalPlan plan) {
             Holder<LocalRelation> optimizedPlan = new Holder<>();
-            plan.forEachDown(Project.class, p -> {
-                List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && (p.child() instanceof EsRelation) == false &&
-                    isNotQueryWithFromClauseAndFilterFoldedToFalse(p)) {
+            var leafRelation = leafRelation(plan);
+
+            // exclude LocalRelations that have been introduced by earlier optimizations (skipped ESRelations)
+            var isNonSkippedLocalRelation = leafRelation instanceof LocalRelation &&
+                ((LocalRelation) leafRelation).executable() instanceof EmptyExecutable == false;
+
+            if (isNonSkippedLocalRelation) {
+                plan.forEachDown(Project.class, p -> {
+                    List<Object> values = extractConstants(p.projections());
+
+                    if (values.size() != p.projections().size()) {
+                        throw new IllegalStateException("Trying to execute non-constant projection in local relation.");
+                    }
+                    if (optimizedPlan.get() != null) {
+                        throw new IllegalStateException("More than one Project found.");
+                    }
+
                     optimizedPlan.set(new LocalRelation(p.source(), new SingletonExecutable(p.output(), values.toArray())));
-                }
-            });
+                });
+            }
 
             if (optimizedPlan.get() != null) {
                 return optimizedPlan.get();
             }
 
             plan.forEachDown(Aggregate.class, a -> {
                 List<Object> values = extractConstants(a.aggregates());
-                if (values.size() == a.aggregates().size() && a.groupings().isEmpty()
-                    && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
+
+                // aggregations on only constant values like "SELECT 'foo' FROM test GROUP BY 1"
```

Comment:
Might be worth adding this statement (or similar) into the optimiser tests too, imo.

Reply:
`testSkipQueryOnEsAggregationWithOnlyConstants` should cover this case